pkg/lib/queueinformer/queueinformer_operator.go

```diff
@@ -1,273 +1,300 @@
+	}
+}
 
-		for _, queueIndexer := range o.queueIndexers {
-			go o.indexerWorker(queueIndexer)
-			go o.indexerWorker(queueIndexer)
+func (o *operator) RunInformers(ctx context.Context) {
+	o.runInformersOnce.Do(func() {
+		o.mu.Lock()
+		defer o.mu.Unlock()
+		for _, informer := range o.informers {
+			go informer.Run(ctx.Done())
 		}
-		ready <- struct{}{}
-		<-stopc
-	}()
 
-	return
+		o.started = true
+		o.logger.Infof("informers started")
+	})
 }
 
-// worker runs a worker thread that just dequeues items, processes them, and marks them done.
-// It enforces that the syncHandler is never invoked concurrently with the same key.
-func (o *Operator) worker(loop *QueueInformer) {
-	for o.processNextWorkItem(loop) {
-	}
+// Run starts the operator's control loops.
+func (o *operator) Run(ctx context.Context) {
+	o.reconcileOnce.Do(func() {
+		go o.run(ctx)
+	})
 }
 
-func (o *Operator) processNextWorkItem(loop *QueueInformer) bool {
-	queue := loop.queue
-	key, quit := queue.Get()
+func (o *operator) run(ctx context.Context) {
+	defer func() {
+		close(o.atLevel)
+		close(o.done)
+	}()
 
-	if quit {
-		return false
+	for _, queueInformer := range o.queueInformers {
+		defer queueInformer.queue.ShutDown()
 	}
-	defer queue.Done(key)
-
-	// requeue five times on error
-	err := o.sync(loop, key.(string))
-	if err != nil && queue.NumRequeues(key.(string)) < 5 {
-		o.Log.Infof("retrying %s", key)
-		utilruntime.HandleError(errors.Wrap(err, fmt.Sprintf("Sync %q failed", key)))
-		queue.AddRateLimited(key)
-		return true
-	}
-	queue.Forget(key)
+
+	errs := make(chan error)
+	go func() {
+		defer close(errs)
+		v, err := o.discovery.ServerVersion()
+		if err != nil {
+			errs <- errors.Wrap(err, "communicating with server failed")
+			return
+		}
+		o.logger.Infof("connection established. cluster-version: %v", v)
+	}()
 
 	select {
-	case o.syncCh <- err:
-	default:
+	case err := <-errs:
+		if err != nil {
+			o.logger.Infof("operator not ready: %s", err.Error())
+			return
+		}
+		o.logger.Info("operator ready")
+	case <-ctx.Done():
+		return
 	}
 
-	if err := loop.HandleMetrics(); err != nil {
-		o.Log.Error(err)
-	}
-	return true
-}
+	o.logger.Info("starting informers...")
+	o.RunInformers(ctx)
 
-func (o *Operator) sync(loop *QueueInformer, key string) error {
-	logger := o.Log.WithField("queue", loop.name).WithField("key", key)
-	obj, exists, err := loop.informer.GetIndexer().GetByKey(key)
-	if err != nil {
-		return err
+	o.logger.Info("waiting for caches to sync...")
+	if ok := cache.WaitForCacheSync(ctx.Done(), o.hasSynced); !ok {
+		o.logger.Info("failed to wait for caches to sync")
+		return
 	}
-	if !exists {
-		// For now, we ignore the case where an object used to exist but no longer does
-		logger.Info("couldn't get from queue")
-		logger.Debugf("have keys: %v", loop.informer.GetIndexer().ListKeys())
-		return nil
+
+	o.logger.Info("starting workers...")
+	for _, queueInformer := range o.queueInformers {
+		go o.worker(ctx, queueInformer)
```

Comment:
the number of workers should probably be configurable?

Reply:
Yup, would a follow up PR be okay?